t/encryption.t

```diff
@@ -11,76 +10,180 @@ BEGIN {
+	    OPCUA::Open62541::Test::Client::planning() + 153;
     } else {
 	plan skip_all => 'open62541 without UA_ENABLE_ENCRYPTION';
     }
 }
 use Test::LeakTrace;
 use Test::NoWarnings;
 
-my %certs;
-@{$certs{server}}{qw(cert key)} = CERT_create(
-    CA => 1,
-    not_after => time() + 365*24*60*60,
-    subject => { commonName => 'OPCUA::Open62541 Test Server' },
-    ext => [{sn => "subjectAltName", data => 'DNS:localhost,URI:urn:open62541.server.application'}],
-);
+require './t/CA.pm';
 
-@{$certs{client}}{qw(cert key)} = CERT_create(
-    CA => 1,
-    not_after => time() + 365*24*60*60,
-    subject => { commonName => 'OPCUA::Open62541 Test Client' },
-    ext => [{sn => "subjectAltName", data => 'DNS:localhost,URI:urn:open62541.client.application'}],
-);
+my $ca = OPCUA::Open62541::Test::CA->new();
+$ca->setup();
 
-# use MIME::Base64 to decode the PEM data into DER
-for my $t (qw(server client)) {
-    for (
-	['cert', \&PEM_cert2string],
-	['key',  \&PEM_key2string],
-    ) {
-	my $pem = $_->[1]->($certs{$t}{$_->[0]});
-	my ($data) = $pem =~ /^-----BEGIN\ [^-]+-----
-	    (.*)
-	    ^-----END\ [^-]+-----/msx;
-	$certs{$t}{"$_->[0]_der"} = decode_base64($data);
-    }
-}
+$ca->create_cert_client(issuer => $ca->create_cert_root(name => "ca_client"));
+$ca->create_cert_server(issuer => $ca->create_cert_root(name => "ca_server"));
+
+$ca->create_cert_server(name => "server_selfsigned");
 
-my $status;
-my $server = OPCUA::Open62541::Test::Server->new(
-    certificate => $certs{server}{cert_der},
-    privateKey  => $certs{server}{key_der},
+$ca->create_cert_server(
+    name        => "server_expired",
+    issuer      => "ca_server",
+    create_args => {not_after => time() - 365*24*60*60},
 );
 
-my $serverconfig = $server->{server}->getConfig();
-$server->start();
+sub _setup {
+    my %args = @_;
+    my $client_name = $args{client_name} // 'client';
+    my $server_name = $args{server_name} // 'server';
 
-my $client = OPCUA::Open62541::Test::Client->new(
-    port => $server->port(),
-    certificate => $certs{client}{cert_der},
-    privateKey  => $certs{client}{key_der},
-);
-my $clientconfig = $client->{client}->getConfig();
-$client->start();
+    my $server = OPCUA::Open62541::Test::Server->new(
+	certificate => $ca->{certs}{$server_name}{cert_pem},
+	privateKey  => $ca->{certs}{$server_name}{key_pem},
+    );
+
+    my $serverconfig = $server->{server}->getConfig();
+    $server->start();
 
-$clientconfig->setSecurityMode(MESSAGESECURITYMODE_SIGNANDENCRYPT);
+    $serverconfig->setApplicationDescription({
+	ApplicationDescription_applicationUri => "urn:server.p5-opcua-open65241",
+	ApplicationDescription_applicationType => APPLICATIONTYPE_SERVER,
+    });
 
-$server->run();
+    my $client = OPCUA::Open62541::Test::Client->new(
+	port => $server->port(),
+	certificate    => $ca->{certs}{$client_name}{cert_pem},
+	privateKey     => $ca->{certs}{$client_name}{key_pem},
+	exists($args{client_trustList}) ? (
```

Comment:
I consider exists not nice in such a case.  Then it is not possible to remove an option by setting the value to undef.  My experience is that using lesser exists and defined makes things simpler.  Just check the boolean value.  If you want to do something with 0 or "", you need defined.  If undef is supposed to be passed explicitly, exists is needed.  But the more of these options are supported, the API gets more complicated.